registry/remote/repository.go

```diff
@@ -361,38 +409,92 @@ func (r *Repository) referrers(ctx context.Context, artifactType string, fn func
+	var index ocispec.Index
 	lr := limitReader(resp.Body, r.MaxMetadataBytes)
-	if err := json.NewDecoder(lr).Decode(&page); err != nil {
+	if err := json.NewDecoder(lr).Decode(&index); err != nil {
 		return "", fmt.Errorf("%s %q: failed to decode response: %w", resp.Request.Method, resp.Request.URL, err)
 	}
-	var refs []ocispec.Descriptor
-	if legacyAPI {
-		refs = page.References
-	} else {
-		refs = page.Referrers
+	referrers := index.Manifests
+	if artifactType != "" &&
+		!isReferrersFilterApplied(index.Annotations[ocispec.AnnotationReferrersFiltersApplied], "artifactType") {
+		// perform client side filtering if the filter is not applied on the
+		// server side
+		referrers = filterReferrers(referrers, artifactType)
 	}
-	// Server may not support filtering. We still need to filter on client side
-	// for sure.
-	refs = filterReferrers(refs, artifactType)
-	if len(refs) > 0 {
-		if err := fn(refs); err != nil {
+	if len(referrers) > 0 {
+		if err := fn(referrers); err != nil {
 			return "", err
 		}
 	}
-
 	return parseLink(resp)
 }
 
+// referrersTagSchema lists the descriptors of manifests directly referencing
+// the given manifest descriptor by requesting referrers tag.
+// fn is called for the referrers result. If artifactType is not empty,
+// only referrers of the same artifact type are fed to fn.
+// reference: https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#backwards-compatibility
+func (r *Repository) referrersTagSchema(ctx context.Context, desc ocispec.Descriptor, artifactType string, fn func(referrers []ocispec.Descriptor) error) error {
+	referrersTag := buildReferrersTag(desc)
+	desc, rc, err := r.FetchReference(ctx, referrersTag)
+	if err != nil {
+		if errors.Is(err, errdef.ErrNotFound) {
+			// no referrers to the manifest
+			return nil
+		} else {
+			return err
+		}
+	}
+	defer rc.Close()
+
+	var index ocispec.Index
+	lr := limitReader(rc, r.MaxMetadataBytes)
+	indexJson, err := content.ReadAll(lr, desc)
+	if err != nil {
+		return fmt.Errorf("failed to read referrers from Referrers tag %s: %w", referrersTag, err)
+	}
+	if err := json.Unmarshal(indexJson, &index); err != nil {
+		return fmt.Errorf("failed to decode referrers from Referrers tag %s: %w", referrersTag, err)
+	}
+
+	referrers := filterReferrers(index.Manifests, artifactType)
+	if len(referrers) == 0 {
+		return nil
+	}
+	return fn(referrers)
+}
+
+// buildReferrersTag builds the referrers tag for the given manifest descriptor.
+// Format: <algorithm>-<digest>
+// Reference: https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#unavailable-referrers-api
+func buildReferrersTag(desc ocispec.Descriptor) string {
+	alg := desc.Digest.Algorithm().String()
+	encoded := desc.Digest.Encoded()
+	return alg + "-" + encoded
+}
+
+// isReferrersFilterApplied checks the filtersApplied annotation to see if
+// filter is in the applied filter list.
+func isReferrersFilterApplied(annotation string, filter string) bool {
```

Comment:
It is more readable to pass the annotation map instead of passing the annotation value.